# pandas I: Series y DataFrames

[![Abrir en Google Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/gf0657-programacionsig/2026-ii/blob/main/contenidos/iii-analisis-visualizacion-datos/12-pandas-series-dataframes.ipynb)

## Trabajo previo

### Lecturas

Antes de la clase, revise las siguientes lecturas, en inglés. El capítulo del libro de McKinney —el creador de pandas— presenta las dos estructuras de datos de este cuaderno; las dos lecciones del curso de Kaggle son breves, con ejercicios interactivos, y cubren la carga de datos y la selección de filas y columnas.

McKinney, W. (2022). Getting started with pandas. En *Python for data analysis: Data wrangling with pandas, NumPy, and Jupyter* (3.ª ed.). O'Reilly Media. https://wesmckinney.com/book/pandas-basics
\
\
Kaggle. (s. f.). Creating, reading and writing. En *Pandas*. Kaggle Learn. Recuperado el 20 de setiembre de 2026, de https://www.kaggle.com/code/residentmario/creating-reading-and-writing
\
\
Kaggle. (s. f.). Indexing, selecting & assigning. En *Pandas*. Kaggle Learn. Recuperado el 20 de setiembre de 2026, de https://www.kaggle.com/code/residentmario/indexing-selecting-assigning

### Otros recursos

La [documentación oficial de pandas](https://pandas.pydata.org/docs/) es la referencia permanente de esta unidad: el tutorial [10 minutes to pandas](https://pandas.pydata.org/docs/user_guide/10min.html) resume lo esencial y la [guía del usuario](https://pandas.pydata.org/docs/user_guide/index.html) desarrolla cada tema. La [hoja de referencia](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf) (*cheat sheet*) oficial condensa en dos páginas las operaciones más usadas.

## Introducción

Este cuaderno abre la unidad de análisis y visualización de datos y, con ella, la serie *pandas*, que continúa con las partes [II: agrupación, uniones y datos faltantes](https://gf0657-programacionsig.github.io/2026-ii/pandas-agrupacion-uniones/) y III (gráficos). La serie *Fundamentos de Python* terminó con la **lista de diccionarios** como representación de una tabla de datos: cada diccionario es una fila y sus claves son las columnas. pandas parte de esa misma idea y la convierte en una estructura especializada, el *DataFrame*, con la que cargar, explorar, filtrar y transformar tablas de datos se resuelve en pocas líneas y sin ciclos.

El hilo conductor de la serie es un conjunto de datos real: los resultados de la [Estimación de Población y Vivienda 2022](https://inec.cr/) del Instituto Nacional de Estadística y Censos (INEC) de Costa Rica, con la población, las viviendas, el área y la densidad de cada provincia, cantón y distrito del país. Son los mismos datos que se usaron, a mano y por unos pocos cantones, en los ejercicios de la serie anterior; ahora se trabajan completos.

Como todos los cuadernos del curso, puede ejecutarse en la nube, con la insignia "Abrir en Google Colab", o localmente, en [Visual Studio Code](https://gf0657-programacionsig.github.io/2026-ii/vscode/) con el ambiente `geopython` como kernel. El archivo `.ipynb` puede descargarse con el botón de descarga (ícono de flecha hacia abajo) en la parte superior de la página del cuaderno, en el sitio web del curso.

## ¿Qué es pandas?

[pandas](https://pandas.pydata.org/) es una biblioteca de Python de código abierto para el **análisis y la manipulación de datos tabulares**. Fue creada por Wes McKinney en 2008, cuando trabajaba en una empresa financiera y necesitaba una herramienta ágil para analizar series de tiempo; su nombre deriva de *panel data*, un término de la econometría. Hoy es la herramienta estándar para trabajar con tablas de datos en Python y la base sobre la que se construyen las bibliotecas geoespaciales de la unidad siguiente: un *GeoDataFrame* de geopandas es un *DataFrame* de pandas con una columna de geometrías.

pandas se apoya en [NumPy](https://numpy.org/), la biblioteca de cómputo numérico de Python, de la que hereda su característica más importante: las operaciones son **vectorizadas**, es decir, se aplican a todos los elementos de una columna a la vez, sin escribir ciclos. Sus dos estructuras de datos son la **Series** (una columna) y el **DataFrame** (una tabla).

Por convención universal, pandas se importa con el alias `pd`:

In [1]:
import pandas as pd

# Versión instalada de pandas
print(pd.__version__)

3.0.5


El ambiente `geopython` del curso tiene pandas 3; Colab puede tener una versión anterior. Las diferencias no afectan a este cuaderno, salvo un detalle de presentación: las columnas de texto se muestran con el tipo `str` en pandas 3 y como `object` en las versiones anteriores.

## Series

Una **Series** es un arreglo unidimensional de valores del mismo tipo acompañado de un **índice**: una etiqueta para cada valor. Puede pensarse como una columna de una tabla o como un cruce entre una lista (valores ordenados) y un diccionario (valores con etiqueta). Se crea a partir de una lista:

In [2]:
# Población de las provincias de Costa Rica
# (INEC, Estimación de Población y Vivienda 2022)
poblacion = pd.Series(
    [1601167, 1035464, 545092, 479117, 412808, 500166, 470383],
    index=["San José", "Alajuela", "Cartago", "Heredia", "Guanacaste", "Puntarenas", "Limón"]
)

poblacion

San José      1601167
Alajuela      1035464
Cartago        545092
Heredia        479117
Guanacaste     412808
Puntarenas     500166
Limón          470383
dtype: int64

A la izquierda se muestra el índice y a la derecha los valores; al final, el tipo de datos (`dtype`) de los valores, en este caso enteros de 64 bits. Si no se especifica un índice, pandas asigna las posiciones 0, 1, 2, ... como en una lista.

Los valores se acceden por su **etiqueta** con `loc[]` o por su **posición** con `iloc[]`:

In [3]:
# Acceso por etiqueta (como en un diccionario)
print(poblacion.loc["Guanacaste"])

# Acceso por posición (como en una lista)
print(poblacion.iloc[0])

412808
1601167


La diferencia con las listas y los diccionarios aparece en las operaciones: una expresión aritmética se aplica a **todos los valores** de la Series a la vez, y el resultado es otra Series con el mismo índice. Esta es la vectorización heredada de NumPy:

In [4]:
# Población en miles de habitantes: la división se aplica a cada valor
poblacion / 1000

San José      1601.167
Alajuela      1035.464
Cartago        545.092
Heredia        479.117
Guanacaste     412.808
Puntarenas     500.166
Limón          470.383
dtype: float64

Compárese con la serie *Fundamentos de Python*, donde el mismo cálculo requería un ciclo `for` que recorriera una lista. Las Series tienen además numerosos métodos estadísticos:

In [5]:
# Estadísticas de la Series
print(poblacion.sum())
print(poblacion.mean())
print(poblacion.max())

# idxmax() retorna la ETIQUETA del valor máximo
print(poblacion.idxmax())

5044197
720599.5714285715
1601167
San José


*Ejercicios de esta sección: ejercicios sobre Series y DataFrames, en la sección de ejercicios al final del cuaderno.*

## DataFrames

Un **DataFrame** es una tabla: un conjunto de Series que comparten el mismo índice, una por columna. Puede crearse directamente a partir de una lista de diccionarios —la estructura con que cerró la serie anterior— en la que cada diccionario se convierte en una fila y cada clave, en una columna:

In [6]:
# Provincias de Costa Rica: población y área
# (INEC, Estimación de Población y Vivienda 2022)
provincias = [
    {"nombre": "San José", "poblacion": 1601167, "area_km2": 4969.80},
    {"nombre": "Alajuela", "poblacion": 1035464, "area_km2": 9772.12},
    {"nombre": "Cartago", "poblacion": 545092, "area_km2": 3093.24},
    {"nombre": "Heredia", "poblacion": 479117, "area_km2": 2663.30},
    {"nombre": "Guanacaste", "poblacion": 412808, "area_km2": 10196.29},
    {"nombre": "Puntarenas", "poblacion": 500166, "area_km2": 11275.00},
    {"nombre": "Limón", "poblacion": 470383, "area_km2": 9176.88}
]

df_provincias = pd.DataFrame(provincias)

df_provincias

,nombre,poblacion,area_km2
0,San José,1601167,4969.80
1,Alajuela,1035464,9772.12
2,Cartago,545092,3093.24
3,Heredia,479117,2663.30
4,Guanacaste,412808,10196.29
5,Puntarenas,500166,11275.00
6,Limón,470383,9176.88


Cuando la última expresión de una celda es un DataFrame, Jupyter lo presenta como una tabla con formato. Sus componentes son el **índice** (a la izquierda, aquí las posiciones 0 a 6), los nombres de las **columnas** (arriba) y los **valores**. Tres atributos los describen:

In [7]:
# Dimensiones: (filas, columnas)
print(df_provincias.shape)

# Nombres de las columnas
print(df_provincias.columns)

# Tipos de datos de cada columna
print(df_provincias.dtypes)

(7, 3)
Index(['nombre', 'poblacion', 'area_km2'], dtype='str')
nombre           str
poblacion      int64
area_km2     float64
dtype: object


Cada columna de un DataFrame es una Series, que se obtiene con el nombre de la columna entre corchetes. Por lo tanto, las operaciones vectorizadas recién vistas funcionan entre columnas, y la densidad de población de las siete provincias se calcula en una sola expresión, sin el ciclo `for` de la serie anterior:

In [8]:
# Una columna es una Series
df_provincias["poblacion"]

0    1601167
1    1035464
2     545092
3     479117
4     412808
5     500166
6     470383
Name: poblacion, dtype: int64

In [9]:
# Densidad de población: operación entre dos columnas completas
df_provincias["poblacion"] / df_provincias["area_km2"]

0    322.179363
1    105.961040
2    176.220403
3    179.895994
4     40.486098
5     44.360621
6     51.257399
dtype: float64

El resultado es una Series nueva; en la sección de columnas nuevas se verá cómo agregarla al DataFrame.

*Ejercicios de esta sección: ejercicios sobre Series y DataFrames, en la sección de ejercicios al final del cuaderno.*

## Carga de datos desde archivos CSV

En la práctica, los DataFrames no se escriben a mano: se cargan desde archivos o servicios web. El formato más común es [**CSV**](https://es.wikipedia.org/wiki/Valores_separados_por_comas) (*comma-separated values*): texto plano con una fila por línea, los valores separados por comas y, usualmente, una primera línea con los nombres de las columnas. Cualquier hoja de cálculo puede exportarlo y casi toda fuente de datos abiertos lo ofrece.

### Los datos del INEC

El INEC publica los resultados de la Estimación de Población y Vivienda 2022 en un [archivo de Excel](https://admin.inec.cr/sites/default/files/2023-11/reResultadosEstimacionPoblacionVivienda2022_3.xlsx) con 12 cuadros. Para el curso, los cuadros 11 y 12 —población, viviendas, área y densidad por provincia, cantón y distrito— se transformaron en tres archivos CSV, uno por nivel, disponibles en el [directorio `datos/inec`](https://github.com/gf0657-programacionsig/2026-ii/tree/main/datos/inec) del repositorio del curso, junto con el archivo original y el programa de Python que hizo la conversión. Las variables se describen en la tabla 1.

<figure style="text-align: center; margin: 20px 0;">
    <figcaption><strong>Tabla 1</strong>. Variables de los archivos <code>cantones-2022.csv</code> y <code>distritos-2022.csv</code>. Fuente: elaboración propia con base en Instituto Nacional de Estadística y Censos (2023).</figcaption>
    <table class="table table-bordered table-striped" style="margin: 0 auto;">
    <thead>
        <tr><th>Variable</th><th>Descripción</th></tr>
    </thead>
    <tbody>
        <tr><td><code>codigo_provincia</code>, <code>provincia</code></td><td>Código (1 a 7) y nombre de la provincia.</td></tr>
        <tr><td><code>codigo_canton</code>, <code>canton</code></td><td>Código (101 a 706) y nombre del cantón, según la división territorial del Censo 2011 (82 cantones).</td></tr>
        <tr><td><code>distrito</code></td><td>Nombre del distrito (solo en el archivo de distritos).</td></tr>
        <tr><td><code>poblacion</code>, <code>hombres</code>, <code>mujeres</code></td><td>Población total y por sexo.</td></tr>
        <tr><td><code>viviendas</code>, <code>viviendas_ocupadas</code>, <code>viviendas_desocupadas</code></td><td>Viviendas individuales, totales y por ocupación.</td></tr>
        <tr><td><code>promedio_ocupantes</code></td><td>Promedio de ocupantes por vivienda ocupada.</td></tr>
        <tr><td><code>area_km2</code></td><td>Área en kilómetros cuadrados.</td></tr>
        <tr><td><code>densidad</code></td><td>Densidad de población (habitantes por km²).</td></tr>
        <tr><td><code>tasa_crecimiento_poblacion</code>, <code>tasa_crecimiento_viviendas</code></td><td>Tasas de crecimiento promedio anual (por cien) entre 2011 y 2022.</td></tr>
        <tr><td><code>relacion_hombre_mujer</code></td><td>Hombres por cada cien mujeres.</td></tr>
    </tbody>
    </table>
</figure>

### read_csv()

La función `pd.read_csv()` lee un archivo CSV y retorna un DataFrame. Acepta tanto una ruta en la computadora local como una URL, lo que permite cargar el archivo directamente desde GitHub —con la URL de su versión "cruda" (*raw*)— y ejecutar el cuaderno en cualquier lugar:

In [10]:
# Carga del archivo CSV de cantones en un DataFrame
cantones = pd.read_csv("https://raw.githubusercontent.com/gf0657-programacionsig/2026-ii/main/datos/inec/cantones-2022.csv")

# Primeras filas
cantones.head()

,codigo_provincia,provincia,codigo_canton,canton,poblacion,hombres,mujeres,viviendas,viviendas_ocupadas,viviendas_desocupadas,promedio_ocupantes,area_km2,densidad,tasa_crecimiento_poblacion,tasa_crecimiento_viviendas,relacion_hombre_mujer
0,1,San José,101,San José,352381,168817,183564,116872,111228,5644,3.16,44.62,7897.4,1.83,2.68,92.0
1,1,San José,102,Escazú,71500,33923,37577,26392,24379,2013,2.92,34.53,2070.7,2.14,3.14,90.3
2,1,San José,103,Desamparados,223226,106209,117017,74568,70380,4188,3.16,118.89,1877.6,0.62,1.96,90.8
3,1,San José,104,Puriscal,38525,18685,19840,14902,13438,1464,2.87,555.04,69.4,1.41,2.76,94.2
4,1,San José,105,Tarrazú,17810,9323,8487,7331,5778,1553,3.07,291.27,61.1,0.82,2.22,109.9


`head()` muestra las primeras cinco filas (o las que se indiquen como argumento) y `tail()`, las últimas. Como el archivo tiene 82 filas y 16 columnas, mostrarlo completo no sería práctico; las primeras filas bastan para confirmar que se cargó bien.

### Exploración inicial

Antes de analizar un conjunto de datos conviene conocer su tamaño, sus tipos de datos y la distribución de sus valores. Los atributos ya vistos y unos pocos métodos resuelven esa exploración:

In [11]:
# Dimensiones
cantones.shape

(82, 16)

In [12]:
# Resumen: columnas, cantidad de valores no nulos y tipos de datos
cantones.info()

<class 'pandas.DataFrame'>
RangeIndex: 82 entries, 0 to 81
Data columns (total 16 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   codigo_provincia            82 non-null     int64  
 1   provincia                   82 non-null     str    
 2   codigo_canton               82 non-null     int64  
 3   canton                      82 non-null     str    
 4   poblacion                   82 non-null     int64  
 5   hombres                     82 non-null     int64  
 6   mujeres                     82 non-null     int64  
 7   viviendas                   82 non-null     int64  
 8   viviendas_ocupadas          82 non-null     int64  
 9   viviendas_desocupadas       82 non-null     int64  
 10  promedio_ocupantes          82 non-null     float64
 11  area_km2                    82 non-null     float64
 12  densidad                    82 non-null     float64
 13  tasa_crecimiento_poblacion  82 non-null     floa

`info()` detecta un problema frecuente: las columnas con **valores faltantes** (menos valores no nulos que filas) y las columnas numéricas que pandas leyó como texto por algún carácter extraño. Aquí no hay ninguno de los dos problemas: todas las columnas tienen 82 valores y los tipos son los esperados (enteros `int64`, decimales `float64` y texto).

In [13]:
# Estadísticas descriptivas de las columnas numéricas
cantones.describe()

,codigo_provincia,codigo_canton,poblacion,hombres,mujeres,viviendas,viviendas_ocupadas,viviendas_desocupadas,promedio_ocupantes,area_km2,densidad,tasa_crecimiento_poblacion,tasa_crecimiento_viviendas,relacion_hombre_mujer
count,82.000000,82.000000,82.000000,82.000000,82.000000,82.000000,82.000000,82.000000,82.000000,82.000000,82.000000,82.000000,82.000000,82.000000
mean,3.402439,347.439024,61514.597561,30632.243902,30882.353659,22390.670732,19772.939024,2617.731707,3.092805,623.739268,792.835366,1.505732,2.794024,101.564634
std,2.023685,200.435335,61667.807726,30175.984828,31553.553209,21099.533599,19538.841018,2266.899406,0.198347,742.353411,1610.753530,0.767684,0.845276,8.219885
min,1.000000,101.000000,6173.000000,3292.000000,2881.000000,3109.000000,2235.000000,267.000000,2.550000,6.750000,14.800000,0.010000,1.210000,82.200000
25%,2.000000,201.250000,23721.500000,11706.000000,11816.500000,9214.250000,7729.000000,1131.250000,2.980000,89.115000,52.400000,0.967500,2.235000,94.700000
50%,3.000000,305.500000,46499.000000,23164.000000,23024.000000,16511.500000,14643.500000,1633.500000,3.090000,280.890000,83.450000,1.460000,2.695000,101.100000
75%,5.000000,507.750000,70057.750000,34427.000000,34873.750000,26479.750000,22565.000000,3395.000000,3.185000,841.745000,711.725000,1.887500,3.192500,107.200000
max,7.000000,706.000000,352381.000000,168817.000000,183564.000000,116872.000000,111228.000000,9449.000000,3.590000,3352.140000,9019.600000,4.230000,6.150000,121.100000


`describe()` calcula, para cada columna numérica, la cantidad de valores, la media, la desviación estándar, el mínimo, los [cuartiles](https://es.wikipedia.org/wiki/Cuartil) y el máximo. Se lee de un vistazo, por ejemplo, que la mitad de los cantones tiene menos de 46 500 habitantes y que la densidad va de 15 a 9020 habitantes por km².

In [14]:
# Distribución de una columna de texto: cantidad de cantones por provincia
cantones["provincia"].value_counts()

provincia
San José      20
Alajuela      16
Guanacaste    11
Puntarenas    11
Heredia       10
Cartago        8
Limón          6
Name: count, dtype: int64

*Ejercicios de esta sección: ejercicios sobre carga y exploración, en la sección de ejercicios al final del cuaderno.*

## Selección de columnas y filas

### Columnas

Ya se vio que una columna se obtiene con su nombre entre corchetes. Para seleccionar **varias columnas** se pasa una **lista** de nombres, es decir, corchetes dentro de corchetes. La distinción es importante: con un nombre se obtiene una Series y con una lista, un DataFrame (aunque la lista tenga un solo nombre):

In [15]:
# Una columna: Series
cantones["canton"].head()

0        San José
1          Escazú
2    Desamparados
3        Puriscal
4         Tarrazú
Name: canton, dtype: str

In [16]:
# Varias columnas: DataFrame (note los corchetes dobles)
cantones[["canton", "poblacion", "densidad"]].head()

,canton,poblacion,densidad
0,San José,352381,7897.4
1,Escazú,71500,2070.7
2,Desamparados,223226,1877.6
3,Puriscal,38525,69.4
4,Tarrazú,17810,61.1


### Filas

Las filas se seleccionan con los mismos accesores de las Series: `loc[]` por etiqueta del índice e `iloc[]` por posición. Como el índice de `cantones` es el de posiciones que asignó `read_csv()`, ambos coinciden en este caso, con una diferencia: en los rangos, `loc[]` **incluye** el límite final e `iloc[]` lo excluye, como el rebanado de las listas:

In [17]:
# Primera fila: una Series cuyo índice son los nombres de las columnas
cantones.iloc[0]

codigo_provincia                     1
provincia                     San José
codigo_canton                      101
canton                        San José
poblacion                       352381
hombres                         168817
mujeres                         183564
viviendas                       116872
viviendas_ocupadas              111228
viviendas_desocupadas             5644
promedio_ocupantes                3.16
area_km2                         44.62
densidad                        7897.4
tasa_crecimiento_poblacion        1.83
tasa_crecimiento_viviendas        2.68
relacion_hombre_mujer             92.0
Name: 0, dtype: object

In [18]:
# Filas en las posiciones 0, 1 y 2 (la 3 se excluye)
cantones.iloc[0:3]

,codigo_provincia,provincia,codigo_canton,canton,poblacion,hombres,mujeres,viviendas,viviendas_ocupadas,viviendas_desocupadas,promedio_ocupantes,area_km2,densidad,tasa_crecimiento_poblacion,tasa_crecimiento_viviendas,relacion_hombre_mujer
0,1,San José,101,San José,352381,168817,183564,116872,111228,5644,3.16,44.62,7897.4,1.83,2.68,92.0
1,1,San José,102,Escazú,71500,33923,37577,26392,24379,2013,2.92,34.53,2070.7,2.14,3.14,90.3
2,1,San José,103,Desamparados,223226,106209,117017,74568,70380,4188,3.16,118.89,1877.6,0.62,1.96,90.8


In [19]:
# Filas con etiquetas 0 a 2 (inclusive) y solo algunas columnas
cantones.loc[0:2, ["canton", "poblacion"]]

,canton,poblacion
0,San José,352381
1,Escazú,71500
2,Desamparados,223226


`loc[]` acepta filas y columnas a la vez, separadas por una coma, y es la forma general de seleccionar en pandas. Se usa con mucha más frecuencia que `iloc[]`, porque las etiquetas son más significativas que las posiciones, y es también la forma de seleccionar filas por condición, como se ve a continuación.

### Filtrado por condiciones

La selección más útil es la **condicional**: quedarse con las filas que cumplen un criterio. Funciona en dos pasos. Primero, una comparación con una columna produce una Series de valores booleanos, uno por fila:

In [20]:
# Una comparación vectorizada produce una Series booleana
cantones["poblacion"] > 100000

0      True
1     False
2      True
3     False
4     False
      ...  
77     True
78    False
79    False
80    False
81    False
Name: poblacion, Length: 82, dtype: bool

Segundo, esa Series booleana entre corchetes selecciona las filas donde el valor es `True`. Es lo que se conoce como **máscara booleana** (*boolean mask*):

In [21]:
# Cantones con más de 100 000 habitantes
cantones[cantones["poblacion"] > 100000]

,codigo_provincia,provincia,codigo_canton,canton,poblacion,hombres,mujeres,viviendas,viviendas_ocupadas,viviendas_desocupadas,promedio_ocupantes,area_km2,densidad,tasa_crecimiento_poblacion,tasa_crecimiento_viviendas,relacion_hombre_mujer
0,1,San José,101,San José,352381,168817,183564,116872,111228,5644,3.16,44.62,7897.4,1.83,2.68,92.0
2,1,San José,103,Desamparados,223226,106209,117017,74568,70380,4188,3.16,118.89,1877.6,0.62,1.96,90.8
7,1,San José,108,Goicoechea,132104,62090,70014,43361,41266,2095,3.19,31.70,4167.3,1.25,2.08,88.7
18,1,San José,119,Pérez Zeledón,156917,80643,76274,63277,54962,8315,2.83,1901.08,82.5,1.40,3.17,105.7
20,2,Alajuela,201,Alajuela,322143,159817,162326,110560,101968,8592,3.08,391.61,822.6,2.13,3.21,98.5
29,2,Alajuela,210,San Carlos,198742,101847,96895,72258,62809,9449,3.15,3352.14,59.3,1.76,3.26,105.1
36,3,Cartago,301,Cartago,165417,80732,84685,53652,49257,4395,3.34,278.66,593.6,1.02,2.41,95.3
44,4,Heredia,401,Heredia,131901,62891,69010,45401,43160,2241,3.05,283.12,465.9,0.59,1.80,91.1
65,6,Puntarenas,601,Puntarenas,141697,71940,69757,55255,45816,9439,3.08,1816.86,78.0,1.90,3.15,103.1
76,7,Limón,701,Limón,116457,61157,55300,40516,34677,5839,3.34,1769.38,65.8,1.91,2.81,110.6


Las condiciones se combinan con los operadores `&` (y), `|` (o) y `~` (no), **cada condición entre paréntesis**. No sirven `and`, `or` y `not`, que operan sobre valores individuales y no sobre Series (la clínica de errores muestra qué ocurre si se usan):

In [22]:
# Cantones de Guanacaste con menos de 30 habitantes por km2
cantones[(cantones["provincia"] == "Guanacaste") & (cantones["densidad"] < 30)]

,codigo_provincia,provincia,codigo_canton,canton,poblacion,hombres,mujeres,viviendas,viviendas_ocupadas,viviendas_desocupadas,promedio_ocupantes,area_km2,densidad,tasa_crecimiento_poblacion,tasa_crecimiento_viviendas,relacion_hombre_mujer
57,5,Guanacaste,504,Bagaces,23709,12602,11107,8798,7545,1253,3.14,1277.93,18.6,1.76,2.71,113.5
62,5,Guanacaste,509,Nandayure,11874,6270,5604,5229,4071,1158,2.91,568.14,20.9,0.60,1.61,111.9
63,5,Guanacaste,510,La Cruz,26829,13793,13036,9443,7786,1657,3.44,1385.40,19.4,3.05,4.77,105.8


In [23]:
# Cantones de Limón o de Puntarenas con más de 50 000 habitantes,
# solo con algunas columnas
cantones.loc[
    (cantones["provincia"].isin(["Limón", "Puntarenas"])) & (cantones["poblacion"] > 50000),
    ["provincia", "canton", "poblacion"]
]

,provincia,canton,poblacion
65,Puntarenas,Puntarenas,141697
67,Puntarenas,Buenos Aires,50562
74,Puntarenas,Corredores,51023
76,Limón,Limón,116457
77,Limón,Pococí,146320
78,Limón,Siquirres,63765


El método `isin()` verifica la pertenencia a una lista de valores y reemplaza una cadena de condiciones unidas con `|`. Por último, la cantidad de filas que cumplen una condición se obtiene con `len()` del resultado o, más directamente, sumando la Series booleana (`True` cuenta como 1):

In [24]:
# ¿Cuántos cantones tienen más de 1000 habitantes por km2?
(cantones["densidad"] > 1000).sum()

np.int64(16)

*Ejercicios de esta sección: ejercicios sobre selección y filtrado, en la sección de ejercicios al final del cuaderno.*

## Ordenamiento

El método `sort_values()` ordena las filas según una o más columnas. Igual que la función `sorted()` de la serie anterior, **retorna un DataFrame nuevo** y deja el original intacto:

In [25]:
# Los cinco cantones más poblados
cantones.sort_values("poblacion", ascending=False).head()

,codigo_provincia,provincia,codigo_canton,canton,poblacion,hombres,mujeres,viviendas,viviendas_ocupadas,viviendas_desocupadas,promedio_ocupantes,area_km2,densidad,tasa_crecimiento_poblacion,tasa_crecimiento_viviendas,relacion_hombre_mujer
0,1,San José,101,San José,352381,168817,183564,116872,111228,5644,3.16,44.62,7897.4,1.83,2.68,92.0
20,2,Alajuela,201,Alajuela,322143,159817,162326,110560,101968,8592,3.08,391.61,822.6,2.13,3.21,98.5
2,1,San José,103,Desamparados,223226,106209,117017,74568,70380,4188,3.16,118.89,1877.6,0.62,1.96,90.8
29,2,Alajuela,210,San Carlos,198742,101847,96895,72258,62809,9449,3.15,3352.14,59.3,1.76,3.26,105.1
36,3,Cartago,301,Cartago,165417,80732,84685,53652,49257,4395,3.34,278.66,593.6,1.02,2.41,95.3


In [26]:
# Ordenamiento por provincia y, dentro de cada provincia, por densidad descendente
cantones.sort_values(["provincia", "densidad"], ascending=[True, False]).head(10)

,codigo_provincia,provincia,codigo_canton,canton,poblacion,hombres,mujeres,viviendas,viviendas_ocupadas,viviendas_desocupadas,promedio_ocupantes,area_km2,densidad,tasa_crecimiento_poblacion,tasa_crecimiento_viviendas,relacion_hombre_mujer
26,2,Alajuela,207,Palmares,39148,18824,20324,13318,12453,865,3.14,38.94,1005.3,1.09,2.36,92.6
20,2,Alajuela,201,Alajuela,322143,159817,162326,110560,101968,8592,3.08,391.61,822.6,2.13,3.21,98.5
22,2,Alajuela,203,Grecia,76226,36848,39378,26746,25136,1610,3.03,142.47,535.0,1.33,2.55,93.6
27,2,Alajuela,208,Poás,36207,18182,18025,12145,10998,1147,3.29,74.48,486.1,1.96,3.14,100.9
25,2,Alajuela,206,Naranjo,46647,22839,23808,16129,14594,1535,3.19,127.03,367.2,0.80,2.06,95.9
24,2,Alajuela,205,Atenas,30407,14788,15619,11257,10086,1171,2.98,126.32,240.7,1.61,2.35,94.7
28,2,Alajuela,209,Orotina,24217,12252,11965,10136,7810,2326,3.09,145.11,166.9,1.59,2.91,102.4
31,2,Alajuela,212,Sarchí,21562,10942,10620,7908,6966,942,3.10,136.77,157.7,1.60,3.10,103.0
30,2,Alajuela,211,Zarcero,15153,7261,7892,5017,4695,322,3.21,157.36,96.3,1.97,3.02,92.0
21,2,Alajuela,202,San Ramón,93264,46321,46943,36416,31246,5170,2.98,1021.75,91.3,1.33,2.76,98.7


Note el **encadenamiento de métodos**: `sort_values()` retorna un DataFrame, sobre el que se invoca `head()`. Este estilo, en el que cada método transforma el resultado del anterior, es característico de pandas y se leerá de izquierda a derecha como una secuencia de pasos.

*Ejercicios de esta sección: ejercicios sobre ordenamiento y columnas nuevas, en la sección de ejercicios al final del cuaderno.*

## Columnas nuevas

Una columna nueva se crea asignando una Series —típicamente, el resultado de una operación entre columnas existentes— a un nombre de columna que no existe. A diferencia de `sort_values()`, la asignación **modifica el DataFrame** en el lugar:

In [27]:
# Porcentaje de mujeres en la población de cada cantón
cantones["porcentaje_mujeres"] = cantones["mujeres"] / cantones["poblacion"] * 100

cantones[["canton", "poblacion", "mujeres", "porcentaje_mujeres"]].head()

,canton,poblacion,mujeres,porcentaje_mujeres
0,San José,352381,183564,52.092479
1,Escazú,71500,37577,52.555245
2,Desamparados,223226,117017,52.420865
3,Puriscal,38525,19840,51.499027
4,Tarrazú,17810,8487,47.653004


Las columnas derivadas sirven también para **verificar** los datos. La densidad que publica el INEC debería coincidir con la población dividida entre el área; el siguiente código la recalcula y comprueba, fila por fila y sin ciclos, que ambas coinciden:

In [28]:
# Densidad recalculada, redondeada a un decimal como la del INEC
cantones["densidad_calculada"] = round(cantones["poblacion"] / cantones["area_km2"], 1)

# ¿Coinciden todas las filas? all() retorna True solo si todos los valores son True
(cantones["densidad_calculada"] == cantones["densidad"]).all()

np.True_

Una columna se elimina con `drop()`, que, como `sort_values()`, retorna un DataFrame nuevo, por lo que el resultado debe asignarse:

In [29]:
# Eliminación de la columna de verificación
cantones = cantones.drop(columns="densidad_calculada")

cantones.columns

Index(['codigo_provincia', 'provincia', 'codigo_canton', 'canton', 'poblacion',
       'hombres', 'mujeres', 'viviendas', 'viviendas_ocupadas',
       'viviendas_desocupadas', 'promedio_ocupantes', 'area_km2', 'densidad',
       'tasa_crecimiento_poblacion', 'tasa_crecimiento_viviendas',
       'relacion_hombre_mujer', 'porcentaje_mujeres'],
      dtype='str')

### Exportación

El camino inverso a `read_csv()` es el método `to_csv()`, que guarda el DataFrame en un archivo. El argumento `index=False` evita que el índice de posiciones se escriba como una columna más:

In [30]:
# Cantones de Heredia ordenados por densidad, guardados en un archivo CSV
heredia = cantones[cantones["provincia"] == "Heredia"].sort_values("densidad", ascending=False)
heredia.to_csv("cantones-heredia.csv", index=False)

# Verificación: se vuelve a leer el archivo recién escrito
pd.read_csv("cantones-heredia.csv").head(3)

,codigo_provincia,provincia,codigo_canton,canton,poblacion,hombres,mujeres,viviendas,viviendas_ocupadas,viviendas_desocupadas,promedio_ocupantes,area_km2,densidad,tasa_crecimiento_poblacion,tasa_crecimiento_viviendas,relacion_hombre_mujer,porcentaje_mujeres
0,4,Heredia,409,San Pablo,29860,14705,15155,10440,9906,534,3.00,8.34,3580.3,0.69,1.76,97.0,50.753516
1,4,Heredia,408,Flores,22026,11067,10959,7467,7200,267,3.06,6.75,3263.1,0.86,1.75,101.0,49.754835
2,4,Heredia,407,Belén,23759,11315,12444,8216,7710,506,3.08,12.38,1919.1,0.85,2.04,90.9,52.375942


En Colab, el archivo se guarda en el almacenamiento temporal de la sesión (visible en el panel de archivos, a la izquierda) y se pierde al cerrarla, a menos que se descargue o se monte Google Drive; en VS Code se guarda en el directorio del cuaderno.

*Ejercicios de esta sección: ejercicios sobre ordenamiento y columnas nuevas, en la sección de ejercicios al final del cuaderno.*

## Clínica de errores

Los errores típicos del trabajo con DataFrames, provocados a propósito para aprender a reconocerlos:

In [31]:
# KeyError: la columna no existe
# (aquí, por una errata en el nombre; la última línea del mensaje muestra el nombre buscado)
cantones["poblacón"]

KeyError: 'poblacón'

In [32]:
# ValueError: "The truth value of a Series is ambiguous"
# (por combinar condiciones con and en vez de &)
cantones[(cantones["provincia"] == "Limón") and (cantones["poblacion"] > 50000)]

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [33]:
# AttributeError: el DataFrame no tiene ese método
# (aquí, por una errata en el nombre del método)
cantones.haed()

AttributeError: 'DataFrame' object has no attribute 'haed'

El segundo error merece atención porque su mensaje es poco intuitivo. `and` intenta decidir si una Series completa es verdadera o falsa, lo que no tiene sentido cuando contiene muchos valores; de ahí lo de "ambiguo". La propia última línea del mensaje sugiere la solución, aunque de manera críptica: usar los operadores vectorizados. Es un buen candidato para pedir una explicación a un asistente de IA con la estructura de prompt de la [serie anterior](https://gf0657-programacionsig.github.io/2026-ii/estructuras-datos-apis/) (rol, contexto, tarea y formato), como propone uno de los ejercicios.

*Ejercicios de esta sección: ejercicios sobre asistentes de IA, en la sección de ejercicios al final del cuaderno.*

## Resumen

- **pandas** es la biblioteca de Python para datos tabulares. Sus operaciones son **vectorizadas**: se aplican a columnas completas, sin ciclos.
- Una **Series** es una columna con **índice**; un **DataFrame** es una tabla de Series con el mismo índice. Un DataFrame se crea desde una **lista de diccionarios** o se carga desde un archivo **CSV** con `pd.read_csv()`, incluso desde una URL.
- Exploración inicial: `head()`, `shape`, `info()`, `describe()` y `value_counts()`.
- Selección: `df["columna"]` (Series), `df[["c1", "c2"]]` (DataFrame), `loc[]` por etiquetas e `iloc[]` por posiciones.
- Filtrado con **máscaras booleanas**: `df[condición]`, con `&`, `|`, `~` y paréntesis para combinar condiciones, e `isin()` para listas de valores.
- `sort_values()` y `drop()` retornan un DataFrame nuevo; la asignación `df["nueva"] = ...` modifica el DataFrame. `to_csv()` exporta.
- Errores típicos: `KeyError` (columna inexistente), `ValueError` por combinar condiciones con `and`/`or` y `AttributeError` (método inexistente).

## Ejercicios

Los ejercicios se agrupan según la sección del cuaderno a la que corresponden; se recomienda realizarlos al concluir la sección respectiva. Resuélvalos en este mismo cuaderno (en su copia de Colab o local). Varios incluyen una **celda de verificación**: ejecútela tal cual después de resolver el ejercicio (en la versión publicada muestra un recordatorio, porque los ejercicios no están resueltos). Las [soluciones de estos ejercicios](https://gf0657-programacionsig.github.io/2026-ii/soluciones-pandas-series-dataframes) se publican después de la clase correspondiente.

### Series y DataFrames

1. Cree una Series llamada `areas` con el área de las siete provincias (los valores de `area_km2` de la lista `provincias` de la sección de DataFrames) y los nombres de las provincias como índice. Use sus métodos para obtener el área total del país y la etiqueta de la provincia más extensa.

In [34]:
# Celda de verificación del ejercicio 1: ejecútela tal cual
try:
    assert len(areas) == 7, "La Series debe tener 7 valores"
    assert "Guanacaste" in areas.index, "El índice debe ser los nombres de las provincias"
    assert round(areas.sum(), 2) == 51146.63, "El área total no coincide con la suma de las siete áreas"
    assert areas.idxmax() == "Puntarenas", "Revise cómo se obtiene la etiqueta del máximo (idxmax)"
    print("Ejercicio 1: ¡correcto!")
except NameError:
    print("Aún no se ha definido la Series areas (ejercicio 1).")

Aún no se ha definido la Series areas (ejercicio 1).


2. **Prediga antes de ejecutar**: ¿qué imprimirá el siguiente programa? Escriba su predicción en una celda de texto y solo después cópielo a una celda de código y ejecútelo.

```python
print(type(cantones["canton"]))
print(type(cantones[["canton"]]))
print(cantones[["canton"]].shape)
```

<details>
<summary>Después de ejecutar, haga clic aquí para ver la explicación</summary>

Imprime <code>&lt;class 'pandas.core.series.Series'&gt;</code>, luego <code>&lt;class 'pandas.core.frame.DataFrame'&gt;</code> y por último <code>(82, 1)</code>. Un nombre entre corchetes simples retorna la columna como Series; una <em>lista</em> de nombres —aunque tenga uno solo— retorna un DataFrame, que conserva la forma de tabla: 82 filas y 1 columna. La distinción importa porque los métodos disponibles y la presentación difieren entre ambos tipos.

</details>

### Carga y exploración

3. Cargue el archivo de distritos (`distritos-2022.csv`, en el mismo directorio del repositorio que el de cantones) en un DataFrame llamado `distritos` y explórelo con `shape`, `info()` y `describe()`. Luego, con `value_counts()` sobre la columna `canton`, determine cuántos distritos tiene el cantón que eligió en la [parte I](https://gf0657-programacionsig.github.io/2026-ii/fundamentos-python/) de la serie *Fundamentos de Python* y cuál es el cantón con más distritos del país.

In [35]:
# Celda de verificación del ejercicio 3: ejecútela tal cual
try:
    assert distritos.shape == (487, 17), "El DataFrame debe tener 487 filas y 17 columnas"
    assert "distrito" in distritos.columns, "Revise que se cargó el archivo de distritos"
    print("Ejercicio 3: ¡correcto!")
except NameError:
    print("Aún no se ha definido el DataFrame distritos (ejercicio 3).")

Aún no se ha definido el DataFrame distritos (ejercicio 3).


### Selección y filtrado

4. Cree un DataFrame llamado `cantones_densos` con los cantones que tienen más de 1000 habitantes por km², con solo las columnas `provincia`, `canton` y `densidad`. ¿A qué provincias pertenecen? Compruébelo con `value_counts()`.

In [36]:
# Celda de verificación del ejercicio 4: ejecútela tal cual
try:
    assert len(cantones_densos) == 16, "Deben quedar 16 cantones; revise la condición"
    assert (cantones_densos["densidad"] > 1000).all(), "Hay cantones con densidad de 1000 o menos"
    assert list(cantones_densos.columns) == ["provincia", "canton", "densidad"], \
        "Deben quedar solo las columnas provincia, canton y densidad, en ese orden"
    print("Ejercicio 4: ¡correcto!")
except NameError:
    print("Aún no se ha definido el DataFrame cantones_densos (ejercicio 4).")

Aún no se ha definido el DataFrame cantones_densos (ejercicio 4).


5. Cree un DataFrame llamado `guanacaste_extensos` con los cantones de Guanacaste que tienen más de 1000 km² de área. Recuerde combinar las dos condiciones con `&` y encerrar cada una entre paréntesis.

In [37]:
# Celda de verificación del ejercicio 5: ejecútela tal cual
try:
    assert set(guanacaste_extensos["canton"]) == {"Liberia", "Nicoya", "Santa Cruz", "Bagaces", "La Cruz"}, \
        "Los cantones no son los esperados; revise las dos condiciones y el operador &"
    print("Ejercicio 5: ¡correcto!")
except NameError:
    print("Aún no se ha definido el DataFrame guanacaste_extensos (ejercicio 5).")

Aún no se ha definido el DataFrame guanacaste_extensos (ejercicio 5).


### Ordenamiento y columnas nuevas

6. Cree una lista de Python llamada `mas_poblados` con los nombres de los cinco cantones más poblados del país, en orden descendente de población. Pista: ordene, seleccione la columna `canton`, tome las primeras cinco filas y convierta el resultado en lista con el método `tolist()`.

In [38]:
# Celda de verificación del ejercicio 6: ejecútela tal cual
try:
    assert type(mas_poblados) == list, "mas_poblados debe ser una lista de Python (use tolist())"
    assert mas_poblados == ["San José", "Alajuela", "Desamparados", "San Carlos", "Cartago"], \
        "Los cantones o su orden no son los esperados; revise ascending"
    print("Ejercicio 6: ¡correcto!")
except NameError:
    print("Aún no se ha definido la lista mas_poblados (ejercicio 6).")

Aún no se ha definido la lista mas_poblados (ejercicio 6).


7. Agregue a `cantones` una columna llamada `porcentaje_desocupadas` con el porcentaje de viviendas desocupadas respecto al total de viviendas de cada cantón. Ordene por esa columna y determine cuál cantón tiene el porcentaje más alto. ¿Qué característica geográfica podría explicarlo?

In [39]:
# Celda de verificación del ejercicio 7: ejecútela tal cual
if "porcentaje_desocupadas" not in cantones.columns:
    print("Aún no se ha creado la columna porcentaje_desocupadas (ejercicio 7).")
else:
    assert round(cantones["porcentaje_desocupadas"].max(), 1) == 44.3, \
        "El valor máximo no es el esperado; revise la fórmula (desocupadas / viviendas * 100)"
    assert cantones.loc[cantones["porcentaje_desocupadas"].idxmax(), "canton"] == "Garabito", \
        "El cantón con el máximo no es el esperado"
    print("Ejercicio 7: ¡correcto!")

Aún no se ha creado la columna porcentaje_desocupadas (ejercicio 7).


8. Construya un DataFrame con los cantones de la provincia del cantón que eligió en la parte I, ordenados de mayor a menor densidad y con solo las columnas `canton`, `poblacion`, `area_km2` y `densidad`. Guárdelo en un archivo llamado `mi-provincia.csv`, sin el índice, y vuelva a cargarlo con `read_csv()` para comprobar que se escribió correctamente.

### Asistentes de IA

9. Pida a un asistente de IA que explique el `ValueError` de la clínica de errores ("The truth value of a Series is ambiguous"), con la estructura de prompt de la serie anterior (rol, contexto, tarea y formato), pegando el código y el mensaje de error completo y pidiendo una explicación y una pista, no el código corregido. Documente en una celda de texto el asistente, el prompt, un resumen de la respuesta y su evaluación de ella: ¿coincide con la explicación de la clínica?, ¿la pista lleva a los operadores `&` y `|`? Esa documentación es la **declaración de uso de IA** que piden los [lineamientos del curso](https://gf0657-programacionsig.github.io/2026-ii/asistentes-ia/).

## Referencias bibliográficas

Instituto Nacional de Estadística y Censos. (2023). *Estimación de población y vivienda 2022: Resultados generales*. INEC. https://admin.inec.cr/sites/default/files/2023-07/rePoblacResultadosGenerales_Estimacion_poblacion_vivienda_2022.pdf
\
\
Instituto Nacional de Estadística y Censos. (2023). *Resultados Estimación de Población y Vivienda 2022* [Conjunto de datos]. INEC. https://admin.inec.cr/sites/default/files/2023-11/reResultadosEstimacionPoblacionVivienda2022_3.xlsx
\
\
Kaggle. (s. f.). Creating, reading and writing. En *Pandas*. Kaggle Learn. Recuperado el 20 de setiembre de 2026, de https://www.kaggle.com/code/residentmario/creating-reading-and-writing
\
\
Kaggle. (s. f.). Indexing, selecting & assigning. En *Pandas*. Kaggle Learn. Recuperado el 20 de setiembre de 2026, de https://www.kaggle.com/code/residentmario/indexing-selecting-assigning
\
\
McKinney, W. (2022). Getting started with pandas. En *Python for data analysis: Data wrangling with pandas, NumPy, and Jupyter* (3.ª ed.). O'Reilly Media. https://wesmckinney.com/book/pandas-basics
\
\
The pandas development team. (s. f.). *pandas documentation*. Recuperado el 20 de setiembre de 2026, de https://pandas.pydata.org/docs/